ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamTestUtils.java

```diff
@@ -76,18 +80,34 @@ static ByteString bytesWritten2ByteString(long bytesWritten) {
 
     @Override
     public CompletableFuture<?> link(DataStream stream, LogEntryProto entry) {
-      final SingleDataStream s = getSingleDataStream(ClientInvocationId.valueOf(entry.getStateMachineLogEntry()));
-      s.setLogEntry(entry);
+      LOG.info("link {}", stream);
+      if (stream == null) {
+        return JavaUtils.completeExceptionally(new IllegalStateException("Null stream: entry=" + entry));
+      }
+      ((SingleDataStream)stream).setLogEntry(entry);
       return CompletableFuture.completedFuture(null);
     }
 
+    @Override
+    public CompletableFuture<Message> applyTransaction(TransactionContext trx) {
```

Comment:
@szetszwo Hi, sorry, why we need to override applyTransaction ? seems no test need this.

Reply:
Yes.  Otherwise, the reply won't have a message.

I just found that there are bugs in both the code and the tests.  The double negative makes the tests passed.  Will fix them.